pkg/runtime/runtime.go

```diff
@@ -695,6 +690,110 @@ func (a *DaprRuntime) beginPubSub(subscribeCtx context.Context, name string, ps
+								return nil
+							}
+							diag.DefaultComponentMonitoring.PubsubIngressEvent(ctx, pubsubName, strings.ToLower(string(pubsub.Retry)), msg.Topic, 0)
+							return err
+						}
+
+						if pubsub.HasExpired(cloudEvent) {
+							log.Debugf("Actor Pubsub: dropping expired pub/sub event %v as of %v", cloudEvent[pubsub.IDField], cloudEvent[pubsub.ExpirationField])
+							diag.DefaultComponentMonitoring.PubsubIngressEvent(ctx, pubsubName, strings.ToLower(string(pubsub.Drop)), msg.Topic, 0)
+							a.sendToDeadLetterIfConfigured(name, msg)
+							return nil
+						}
+						// Actor calling if ActorType is correct
+						if cloudEvent[pubsub.ActorTypeField] == "" || cloudEvent[pubsub.ActorTypeField] == actorPubSubSubscription.ActorType {
+							err := a.callActorPubsub(ctx, cloudEvent, actorPubSubSubscription)
+							if err != nil {
+								return err
+							}
+							return nil
+						}
+					}
+				}
+				return nil
+			}); subscribeError != nil {
+				log.Error("Actor Pubsub: Failed to subscribe to topic %s", actorConfigSubscriptions.Topic)
+				return subscribeError
+			}
+		}
+	}
+	return nil
+}
+
+// actorCallPubsub calls the actor that was declared in the cloudevent or in the app.Config
+func (a *DaprRuntime) callActorPubsub(ctx context.Context, cloudEvent map[string]interface{}, actorPubSubSubscription config.PubSubConfig) error {
+	sendData, _ := json.Marshal(cloudEvent[pubsub.DataField])
+	if cloudEvent[pubsub.ActorIdField] == "" {
+		// Actor id is not in the cloudevent
+		if actorPubSubSubscription.ActorIDDataAttribute == "" {
+			log.Debug("Actor Pubsub: ActorID and ActorIDDataAttribute not declared in cloud event id - %s", cloudEvent[pubsub.IDField])
+		} else {
+			dataMap := make(map[string]string)
+			// Must be a string string map
+			err := json.Unmarshal(sendData, &dataMap)
+			if err != nil {
+				log.Debugf("Actor Pubsub: Error Actor ID attribute can only be used with JSON type data: %s", err)
+				return err
+			}
+			if attribute, ok := dataMap[actorPubSubSubscription.ActorIDDataAttribute]; ok {
+				req := invokev1.NewInvokeMethodRequest(actorPubSubSubscription.Method)
+				req.WithActor(actorPubSubSubscription.ActorType, attribute)
+				req.WithHTTPExtension(nethttp.MethodPost, "")
+				req.WithRawData(sendData, contenttype.CloudEventContentType)
+				log.Infof("Invoking Actor %+v: with method %s", req.Actor(), req.Message().GetMethod())
+
+				res, err := a.actor.Call(ctx, req)
+				if err != nil {
+					log.Debugf("Error invoking actor: %s", err.Error())
+					return err
+				}
+				log.Debugf("Actor response: %v", res.Message().GetData())
+			} else {
+				log.Debug("Actor Pubsub: Not a valid ActorIDAttribute for this actor")
+			}
+		}
+	} else {
+		// actor id is in the cloud event
+		actorid := fmt.Sprint(cloudEvent[pubsub.ActorIdField])
+		req := invokev1.NewInvokeMethodRequest(actorPubSubSubscription.Method)
```

Comment:
This code block seems to be very similar to the invocation above. Can this be refactored so we reuse the actor invocation logic?